Documentation/Relationship/Relationship.md

```diff
@@ -0,0 +1,333 @@
+
+```json
+{
+  "id": "472",
+  "name": "Andreas",
+  "lastname": "Bauer",
+  "email": "mail@example.org",
+  "_links": {
+    "self": "https://example.api/user/472"
+  }
+}
+```
+
+_Note: to be fair the `self` link can also be more easily retrieved by looking at what the client sent in its request.
+However what we actually do is building a mapping from types to their respective `Handler`
+which is needed for the following chapters._
+
+### 3.2. `Handler`s which return an `Identifiable` but don't have a parameter
+
+A special case to [3.1](#31-looking-at-the-return-type-of-handle) is, when we detect that a `Handler` returns a type
+conforming to `Identifiable`, but doesn't have a `@PathParameter` which can be mapped to `.id` property.
+
+This case is illustrated by the following example web service with the `MeUserHandler`:
+```swift
+struct TestService: WebService, Component {
+  @PathParameter var userId: User.ID
+
+  var content: some Component {
+    Group("user", $postId) {
+      UserHandler(userId: $userId)
+      // could contain other routes giving more information for the given user, e.g. ./posts
+    }
+    Group("me") {
+      MeUserHandler() // handle() returns `User`
+    }
+  }
+}
+```
+
+In such a case, we can search the web service for a `Handler` which also returns `User`, but has a `@PathParameter`.
+We could then find the `UserHandler` registered under `/user/:userId`.
+
+A response returned from the `MeUserHandler` would then include a `self` relationship pointing to `/user/{userId}` 
+(`userId` is known from the returned `User` instance). More specifically the response will include all relationships
+which are also returned on a request to `user/:userId` (meaning links to potential subroutes).
+
+**Open questions:**
+* This approach heavily relies on the fact that `/me` returns the same type as `user/:userId`. It might be worth
+  creating a explicit mechanism to define the `self` link in a case were `/me` would only return a subset of the properties
+  of `User`.  
+  This could be solved using the same mechanism as presented in [3.3](#33-defining-foreignkeys). 
+
+### 3.3. Defining `ForeignKeys`
+
+In this chapter we assume the existence of the example webservice from the previous chapter and adding a `/article`
+handler.
+
+```swift
+var content: some Component {
+  // ...
+  Group("article", $articleId) {
+    ArticleHandler(articleId: $articleId)
+  }
+}
+```
+
+The idea is, that every `Article` is written by a certain `User` (aka the author).
+A potential REST API response should then include a link to the respective `User` resource.  
+To be able to display such a relationship the user needs to provide that information to us.
+
+A `Article` could then be defined like the following:
+```swift
+struct Article: IdentifiableWithForeignKeys {
+  var id: String
+  var heading: String
+  var content: String
+
+  var writtenBy: String
+
+  static var foreignKeys: some ForeignKeys {
```

Comment:
Shouldn't we use just an array of User parameters? Like
`var writers : [User]`
Correct me if I am wrong.
I think the GraphQL usage of it will be as in the following
query {
 article(name: "Something") {
writer(name: "One of the writer"){
name
surname
}
content
}
}

Reply:
The idea in this **example** is that every `Article` has one author (in the example a `User`). As `User`s are handled by the Handler registered under `/user/:userId` we want to basically "link" to that endpoint. With the above declaration we give Apodini the hint on what the `userId` is and where to find it.
In REST this would result in a `writtenBy` link in the `_links` section.
In GraphQL you could do a query like the following (if I'm not wrong)
```graphql
query {
  article(id: "1237) {
    writtenBy {
      name
      lastname
      email
    }
  }
}
```
----
(@nerdsupremacist)

This basically models 1:n relationships.
But as you suggest, we probably also need to think about n:m relationships, meaning e.g. Articles which are written by multiple authors.

I could see us adding a `ForeignKey` implementation which accepts a KeyPath to a variable holding an array of identifiers.

Assuming the Article example now has `var writtenBy: [String]` declared (with some proper way of defining that as ForeignKey).
I could envision that the REST endpoint would just insert an array of links to all known authors 
```json5
{
  // ...
  "_links": {
    "writtenBy": ["https://example.api/users/123", "https://example.api/users/124"],
    "self": "..."
}
```

And for the GraphQL exporter a query like the following would be possible:
```graphql 
query {
  article(id: "1237) {
    writtenBy {
      edges {
        nodes {
          name
          lastname
          email
        }
      }
    }
  }
}
```

Reply:
Something like `var writers : [User]` would be intuitive and easy, but it doesn't scale well.

For example think of a Post <-> Author relationship:

```swift
struct Post {
    let text: String
    let author: User
}

struct User {
    let name: String
    let posts: [Post]
}
```

results in an endless loop. Plus usually you don't load everything into memory at once, since loading the posts for a user or vice versa could be costly (e.g. loading from a DB). In these cases keeping track of the ids is simpler

Reply:
In the case of N:M relationships I think that we can for now model it via additional nested groups

```swift
Group("article", $articleID) {
    ArticleHandler(id: $articleID)
    Group("authors") {
        AuthorsOfArticles(id: $articleID)
    }
}
```

which for REST would lead to

/article/:id -> { ..., "_links": { "self": "/article/:id", "authors": "/article/:id/authors" } } 
/article/:id/authors

And for GraphQL:

```
query {
    article(id: 123) {
        authors {
            name
            email
        }
    }
}
```